<a href="https://colab.research.google.com/github/StephanieO-py/New-York-Bus-Breakdown-Analysis/blob/main/New_York_Bus_Breakdown_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# AfterWork Data Science: Data Wrangling with Python Project

## 1. Defining the Question
The dataset you have provided comes from the Bus Breakdown and Delay system, which
collects information from school bus vendors operating in the field in real-time. Bus staff that
encounter delays during the route are instructed to radio the dispatcher at the bus vendor’s
central office. The bus vendor staff log into the Bus Breakdown and Delay system to record the
event and notify OPT. OPT customer service agents use this system to inform parents who call
with questions regarding bus service. The Bus Breakdown and Delay system is publicly
accessible and contains real-time updates. All information in the system is by school bus vendor
staff




>> Objectives
1. Read
2. Explore
3. Clean
4. Perform EDA




 








### a) Specifying the Data Analysis Question

Specify the reasearch question that you'll be answering.

*   Which bus companies had the highest breakdowns?
*   What were the top 3 reasons for bus delays?
*   How many students were on the buses when they broke down?
*   Which were the most frequent reasons for bus breakdowns?
*   What were the most frequent reasons for the bus running late?
*   What was the average delay time of each reason type?

### b) Defining the Metric for Success

How will you know that your solution will have satisfied your research question?

### c) Understanding the context 

The Data is owned by NYC Open Data and is provided by the Deapartment of Education


Dataset Source:
https://data.cityofnewyork.us/Transportation/Bus-Breakdown-and-Delays/ez4e-fazm



### d) Recording the Experimental Design

Describe the steps/approach that you will use to answer the given question.

### e) Data Relevance

How relevant was the provided data?

## 2. Reading the Data

In [ ]:
# Importing our libraries
# ---
#
import pandas as pd # for data maniuplation
import numpy as np # for scientific computing

In [ ]:
# Load the data below
# --- 
# Dataset url = https://bit.ly/BusBreakdownDataset
# --- 
# 

Bus_Breakdown = pd.read_csv('https://bit.ly/BusBreakdownDataset')
Bus_Breakdown

School_Year  ...  School_Age_or_PreK
0        2015-2016  ...          School-Age
1        2015-2016  ...          School-Age
2        2015-2016  ...               Pre-K
3        2015-2016  ...          School-Age
4        2015-2016  ...          School-Age
...            ...  ...                 ...
281105   2016-2017  ...               Pre-K
281106   2016-2017  ...               Pre-K
281107   2016-2017  ...          School-Age
281108   2016-2017  ...          School-Age
281109   2016-2017  ...          School-Age

[281110 rows x 21 columns]

In [ ]:
# Checking the first 5 rows of data
# ---
#
Bus_Breakdown.head(5)

School_Year  Busbreakdown_ID  ... Breakdown_or_Running_Late School_Age_or_PreK
0   2015-2016          1227538  ...              Running Late         School-Age
1   2015-2016          1227539  ...              Running Late         School-Age
2   2015-2016          1227540  ...              Running Late              Pre-K
3   2015-2016          1227541  ...              Running Late         School-Age
4   2015-2016          1227542  ...              Running Late         School-Age

[5 rows x 21 columns]

In [ ]:
# Checking the last 5 rows of data
# ---
#
Bus_Breakdown.tail(5)

School_Year  ...  School_Age_or_PreK
281105   2016-2017  ...               Pre-K
281106   2016-2017  ...               Pre-K
281107   2016-2017  ...          School-Age
281108   2016-2017  ...          School-Age
281109   2016-2017  ...          School-Age

[5 rows x 21 columns]

In [ ]:
# Sample 10 rows of data
# ---
#
Bus_Breakdown.sample(10)

School_Year  ...  School_Age_or_PreK
271178   2018-2019  ...          School-Age
111842   2017-2018  ...          School-Age
135999   2018-2019  ...          School-Age
18071    2015-2016  ...          School-Age
67736    2016-2017  ...          School-Age
34435    2015-2016  ...          School-Age
101447   2016-2017  ...          School-Age
172491   2017-2018  ...          School-Age
258346   2018-2019  ...          School-Age
91341    2016-2017  ...          School-Age

[10 rows x 21 columns]

In [ ]:
# Checking number of rows and columns
# ---
#  
Bus_Breakdown.shape

(281110, 21)

In [ ]:
# Checking datatypes
# ---
# 
Bus_Breakdown.dtypes

School_Year                        object
Busbreakdown_ID                     int64
Run_Type                           object
Bus_No                             object
Route_Number                       object
Reason                             object
Schools_Serviced                   object
Occurred_On                        object
Created_On                         object
Boro                               object
Bus_Company_Name                   object
How_Long_Delayed                   object
Number_Of_Students_On_The_Bus       int64
Has_Contractor_Notified_Schools    object
Has_Contractor_Notified_Parents    object
Have_You_Alerted_OPT               object
Informed_On                        object
Incident_Number                    object
Last_Updated_On                    object
Breakdown_or_Running_Late          object
School_Age_or_PreK                 object
dtype: object

Record your observations below:

*   The dataset contains 21 columns and 281110 rows.
*   19 columns have are of the string data type while 2 are integers. However, School_Year, Occured_On, Created_On, Informed_Number and Last_Updated_On should be of the Date-Time data type



## 3. External Data Source Validation

You can make sure your data matches other credible information. This ensures that the measurements are roughly in line with what they should be and it serves as a check on what other things might be wrong in your dataset.

External validation can often be as simple as checking your data against some other data i.e. country population your dataset with country population on world bank data.

Replace the text found on this text cell with your external data source validation information.

## 4. Data Preparation

### Performing Data Cleaning

In [ ]:
# Checking datatypes and missing entries of all the variables
# ---
# 

Bus_Breakdown.isnull().any()

School_Year                        False
Busbreakdown_ID                    False
Run_Type                            True
Bus_No                              True
Route_Number                        True
Reason                              True
Schools_Serviced                    True
Occurred_On                        False
Created_On                         False
Boro                                True
Bus_Company_Name                   False
How_Long_Delayed                    True
Number_Of_Students_On_The_Bus      False
Has_Contractor_Notified_Schools    False
Has_Contractor_Notified_Parents    False
Have_You_Alerted_OPT               False
Informed_On                        False
Incident_Number                     True
Last_Updated_On                    False
Breakdown_or_Running_Late          False
School_Age_or_PreK                 False
dtype: bool

We observe the following from our dataset:

*   8 columns out of the 21 have missing data



In [ ]:
# Standardizing your dataset i.e. variable renaming
# ---
# 
Bus_Breakdown.columns = Bus_Breakdown.columns.str.strip().str.lower()
Bus_Breakdown.head()

school_year  busbreakdown_id  ... breakdown_or_running_late school_age_or_prek
0   2015-2016          1227538  ...              Running Late         School-Age
1   2015-2016          1227539  ...              Running Late         School-Age
2   2015-2016          1227540  ...              Running Late              Pre-K
3   2015-2016          1227541  ...              Running Late         School-Age
4   2015-2016          1227542  ...              Running Late         School-Age

[5 rows x 21 columns]

We observe the following from our dataset:

*   Changing the column names to lower case makes the dataset look neater and easier to read.



In [ ]:
# Checking how many duplicate rows are there in the data
# ---
# 

sum(Bus_Breakdown.duplicated())


0

We observe the following from our dataset:

*   The dataset has no duplicated data



In [ ]:
# Checking if any of the columns are all null
# ---
# 
Bus_Breakdown_without_columns = Bus_Breakdown.dropna(axis=1, how='all')

Bus_Breakdown_without_columns.shape

(281110, 21)

We observe the following from our dataset:

*   No column was dropped, indicating that there is no column that entirely contained null values


In [ ]:
# Checking if any of the rows are all null
# ---
#
Bus_Breakdown_cleaned = Bus_Breakdown_without_columns.dropna(how='all')

Bus_Breakdown_cleaned.shape

(281110, 21)

We observe the following from our dataset:

*   No row was dropped, indicating that there is no row that entirely caontained null values



In [ ]:
# Checking if the "Yes/No" fields contain only these 2 values
# for have_you_alerted_opt variable
# ---
# Hint: Use unique() function
#

Bus_Breakdown_cleaned.have_you_alerted_opt.unique()

array(['Yes', 'No'], dtype=object)

We observe the following from our dataset:

*  The 'have_you_alerted_op' variable only contains 'Yes' and 'No' values



In [ ]:
# Checking if the "Yes/No" fields contain only these 2 values
# for has_contractor_notified_parents variable
# ---
# 
Bus_Breakdown_cleaned.has_contractor_notified_parents.unique()

array(['No', 'Yes'], dtype=object)

We observe the following from our dataset:

*   The 'has_contractor_notified_parents' variable only contains 'Yes' and 'No' values



In [ ]:
# Checking if the "Yes/No" fields contain only these 2 values
# for has_contractor_notified_schools variable
# ---
#  
Bus_Breakdown_cleaned.has_contractor_notified_schools.unique()

array(['Yes', 'No'], dtype=object)

We observe the following from our dataset:

*   The 'has_contractor_notified_schools' variable only contains 'Yes' and 'No' values



In [ ]:
# Checking unique values in break_down_or_running_late variable to ensure there is no duplication
# ---
#

bus_breakdown_df = Bus_Breakdown_cleaned[Bus_Breakdown_cleaned.duplicated(['breakdown_or_running_late'])]
bus_breakdown_df.shape



(281108, 21)

In [ ]:
# Checking unique values in school_age_or_prek variable
# ---
# 

bus_breakdown = bus_breakdown_df[bus_breakdown_df.duplicated(['school_age_or_prek'])]
bus_breakdown.shape

(281106, 21)

In [ ]:
# Checking unique values in school_year variable
# ---
#  
new_bus_breakdown = bus_breakdown[bus_breakdown.duplicated(['school_year'])]
new_bus_breakdown.shape


(281101, 21)

In [ ]:
# Checking unique values in reason variable
# ---
#  
reason_for_breakdown = new_bus_breakdown[new_bus_breakdown.duplicated(['reason'])]
reason_for_breakdown.shape

(281090, 21)

In [ ]:
# Checking unique values in run_type variable
# ---
#  
unique_bus_breakdown = reason_for_breakdown[reason_for_breakdown.duplicated(['run_type'])]
unique_bus_breakdown.shape

(281079, 21)

In [ ]:
# Checking unique values in boro variable
# ---
#  

final_bus_breakdown = unique_bus_breakdown[unique_bus_breakdown.duplicated(['boro'])]
final_bus_breakdown.shape

(281067, 21)

### Overall Data Cleaning Observations
**Missing Values**

- There are a large number of missing values in the fields "How_Long_Delayed" which is important to our analysis. 
- There is an extremely large number of missing values in the "Incident_number" field but this is not incidental to our analysis and cannot be filled in without additional information.

**Error in values**

- "How_Long_Delayed" contains string values such as "MINS" or "mins" and a range of values, which needs to be changed to single integer value for our analysis.

**Error in Datatypes**

- "How_Long_Delayed" is a string datatype, should be converted to integer type.

**Error in field names**
- The column name "Boro" should be renamed to "Borough".


### Next Steps: Data Cleaning Steps

**Error in values**

- Extract the first integer value (lowest delay time) in the column "How_Long_Delayed"


**Missing Values**

- Impute the missing values in the field "How_Long_Delayed" with the mean value.


**Error in Datatypes**

- Convert "How_Long_Delayed" to int datatype.



**Error in field names**

- Rename the column "Boro" to "Borough".

In [ ]:
# Lets first start by creating a copy of our dataframe 
# df_clean = df.copy(). We will use this copy as our cleaning copy.
# ---
#
df_clean = final_bus_breakdown.copy()
df_clean.head()

school_year  busbreakdown_id  ... breakdown_or_running_late school_age_or_prek
8    2015-2016          1227546  ...              Running Late         School-Age
9    2015-2016          1227547  ...              Running Late         School-Age
10   2015-2016          1227551  ...              Running Late         School-Age
15   2015-2016          1227557  ...              Running Late         School-Age
19   2015-2016          1227564  ...              Running Late              Pre-K

[5 rows x 21 columns]

In [ ]:
# Then extracting the lowest delay time in the column how_long_delayed from the string
# 
df_clean['how_long_delayed'] = df_clean['how_long_delayed'].str.extract('(\d+)')
df_clean['how_long_delayed'].head()
df_clean.head()

school_year  busbreakdown_id  ... breakdown_or_running_late school_age_or_prek
8    2015-2016          1227546  ...              Running Late         School-Age
9    2015-2016          1227547  ...              Running Late         School-Age
10   2015-2016          1227551  ...              Running Late         School-Age
15   2015-2016          1227557  ...              Running Late         School-Age
19   2015-2016          1227564  ...              Running Late              Pre-K

[5 rows x 21 columns]

We impute the null values in 'how_long_delayed' column with mean of the column. This will take a couple of steps...

In [ ]:
# We first convert our how_long_delayed to float type to allow for imputation
# ---
# 

df_clean['how_long_delayed'] = df_clean['how_long_delayed'].astype('float')
df_clean['how_long_delayed'].dtypes

dtype('float64')

In [ ]:
# Then later perform our mean imputation 
# ---
# 

mean= df_clean['how_long_delayed'].mean()
round(mean,2)
df_clean['how_long_delayed'] = df_clean['how_long_delayed'].replace(np.nan, 28.15)
df_clean.head(15)

school_year  busbreakdown_id  ... breakdown_or_running_late school_age_or_prek
8    2015-2016          1227546  ...              Running Late         School-Age
9    2015-2016          1227547  ...              Running Late         School-Age
10   2015-2016          1227551  ...              Running Late         School-Age
15   2015-2016          1227557  ...              Running Late         School-Age
19   2015-2016          1227564  ...              Running Late              Pre-K
23   2015-2016          1227394  ...              Running Late              Pre-K
24   2015-2016          1227566  ...              Running Late         School-Age
26   2015-2016          1227570  ...              Running Late              Pre-K
27   2015-2016          1227571  ...              Running Late         School-Age
28   2015-2016          1227574  ...              Running Late         School-Age
29   2015-2016          1227575  ...              Running Late         School-Age
31   2015-2016          1227579  ...              Running Late              Pre-K
32   2015-2016          1227580  ...              Running Late         School-Age
33   2015-2016          1227582  ...              Running Late         School-Age
34   2015-2016          1227583  ...              Running Late         School-Age

[15 rows x 21 columns]

In [ ]:
# Then convert back our how_long_delayed column to integer datatype
# ---
# 
df_clean['how_long_delayed'] = df_clean['how_long_delayed'].astype('int')
df_clean['how_long_delayed'].dtypes

dtype('int64')

In [ ]:
# Then check for nulls in the column
# ---
# 
df_clean['how_long_delayed'].isnull().any()

False

In [ ]:
# Rename Boro column to Borough
# ---
# 
df_clean.rename(columns = {'boro':'borough'}, inplace = True)
df_clean

school_year  ...  school_age_or_prek
8        2015-2016  ...          School-Age
9        2015-2016  ...          School-Age
10       2015-2016  ...          School-Age
15       2015-2016  ...          School-Age
19       2015-2016  ...               Pre-K
...            ...  ...                 ...
281105   2016-2017  ...               Pre-K
281106   2016-2017  ...               Pre-K
281107   2016-2017  ...          School-Age
281108   2016-2017  ...          School-Age
281109   2016-2017  ...          School-Age

[281067 rows x 21 columns]

In [ ]:
# Lastly we convert all values in our columns to lower case
# for ease of reading
# ---
# 
# 

#for columns in df_clean.columns[x]:
       #df_clean[columns] = df_clean[columns].str.lower()

df_clean['run_type']=df_clean['run_type'].str.lower()
df_clean['reason']=df_clean['reason'].str.lower()
df_clean['borough']=df_clean['borough'].str.lower()
df_clean['bus_company_name']=df_clean['bus_company_name'].str.lower()
df_clean['has_contractor_notified_schools']=df_clean['has_contractor_notified_schools'].str.lower()
df_clean['has_contractor_notified_parents']=df_clean['has_contractor_notified_parents'].str.lower()
df_clean['have_you_alerted_opt']=df_clean['have_you_alerted_opt'].str.lower()
df_clean['breakdown_or_running_late']=df_clean['breakdown_or_running_late'].str.lower()
df_clean['school_age_or_prek']=df_clean['school_age_or_prek'].str.lower()


In [ ]:
# Check the first 5 record the cleaned dataset
# ---
#
df_clean.head(5)

school_year  busbreakdown_id  ... breakdown_or_running_late school_age_or_prek
8    2015-2016          1227546  ...              running late         school-age
9    2015-2016          1227547  ...              running late         school-age
10   2015-2016          1227551  ...              running late         school-age
15   2015-2016          1227557  ...              running late         school-age
19   2015-2016          1227564  ...              running late              pre-k

[5 rows x 21 columns]


## 5. Solution Implementation

Here we investigate the questions that would help craft our recommendations.

### 5.a) Questions

In [ ]:
# 1. Which bus companies that had the highest breakdowns?
# ---
#
breakdowns = df_clean.groupby(['bus_company_name']).count()

# Sort to get bus company with highest breakdowns
# ---
# YOUR CODE GOES BELOW
#

breakdowns.sort_values(by='bus_company_name', ascending=1)
breakdowns.head(10)

school_year  ...  school_age_or_prek
bus_company_name                            ...                    
1967                                     1  ...                   1
1992                                   108  ...                 108
`                                        1  ...                   1
acme bus corp. (b2321)                1601  ...                1601
addies                                  26  ...                  26
alina                                    1  ...                   1
alina services corp.                   534  ...                 534
all american school bus c             1268  ...                1268
all american school bus corp.          758  ...                 758
all county bus llc (b2321              255  ...                 255

[10 rows x 20 columns]

In [ ]:
# 2. What were the top 3 reasons for bus delays?
# ---
#
bus_delay = df_clean.groupby(['reason']).count()[['how_long_delayed']].reset_index()

# Sort to get the most frequent reason
# ---
# YOUR CODE GOES BELOW
#

bus_delay.sort_values(by='reason', ascending = 0)
bus_delay.head(3)

reason  how_long_delayed
0           accident              2525
1  delayed by school              2221
2          flat tire              8408

In [ ]:
# 3. How many students were in the buses when they broke down? 
# ---
#
students=df_clean.groupby(['number_of_students_on_the_bus']).count()[['busbreakdown_id']].reset_index()
students.sum()

number_of_students_on_the_bus    310996
busbreakdown_id                  281067
dtype: int64

In [ ]:
# 4. Which were most frequent reasons for bus breakdowns?
# ---
#
breakdown_reasons = df_clean[df_clean.breakdown_or_running_late == 'breakdown'].groupby(['reason']).count()

# Sort to get most frequent reasons
# ---
# YOUR CODE GOES BELOW
#

breakdown_reasons.sort_values(by='reason', ascending = 0)
breakdown_reasons.head

<bound method NDFrame.head of                              school_year  ...  school_age_or_prek
reason                                    ...                    
accident                             202  ...                 202
delayed by school                      7  ...                   7
flat tire                           4037  ...                4037
heavy traffic                        419  ...                 419
late return from field trip           35  ...                  35
mechanical problem                 14985  ...               14985
other                               3528  ...                3528
problem run                           28  ...                  28
weather conditions                    82  ...                  82
won`t start                         7729  ...                7729

[10 rows x 20 columns]>

In [ ]:
# 5. What were the most frequent reasons for the bus running late? 
# ---
#
reasons = df_clean.groupby(['reason','how_long_delayed']).count()

# Get the records with running late reasons and sort to get most frequent reasons
# ---
# YOUR CODE GOES BELOW
#

reasons.sort_values(by ='reason', ascending = 0)
reasons

school_year  ...  school_age_or_prek
reason      how_long_delayed               ...                    
accident    0                          32  ...                  32
            1                         157  ...                 157
            2                           6  ...                   6
            3                           1  ...                   1
            5                           3  ...                   3
...                                   ...  ...                 ...
won`t start 160                         1  ...                   1
            627                         1  ...                   1
            3256                        1  ...                   1
            4032                        1  ...                   1
            5594                        1  ...                   1

[491 rows x 19 columns]

In [ ]:
# 6. What was the average delay time for each reason type?
# ---
# 
avg_delay = df_clean.groupby('reason').mean().reset_index()

# Get the records with reasons and how long on average a delay took then sort 
# ---
# YOUR CODE GOES BELOW
#

avg_delay.sort_values(by='reason', ascending = 0)
avg_delay

reason  ...  number_of_students_on_the_bus
0                     accident  ...                       6.305347
1            delayed by school  ...                       5.884737
2                    flat tire  ...                       1.720861
3                heavy traffic  ...                       3.926396
4  late return from field trip  ...                       4.395125
5           mechanical problem  ...                       1.552486
6                        other  ...                       2.738197
7                  problem run  ...                       6.408655
8           weather conditions  ...                       2.378012
9                  won`t start  ...                       0.919958

[10 rows x 4 columns]

### 5.b) Recommendations

From the above analysis, below are our recommendations:

1.   Since the second most frequent reason for bus delay is the school, they should implement a system where the school will keep time. The teachers should release the students on time rather than 5 minutes later
 





## 6. Challenging your Solution

During this step, we review our solution and implement  approaches that could potentially provide a better outcome. In our case, we could propose the following question that wasn't answered in our solution because it couldn't have greatly contributed to our recommendation.

In [ ]:
# Which boroughs experienced the most breakdowns?
# ---  
#
breakdowns = df_clean.groupby(['borough']).count()

# Sort to get bus company with highest breakdowns
# ---
# 
#

breakdowns.sort_values(by = 'borough', ascending =0)
breakdowns.head

<bound method NDFrame.head of                  school_year  ...  school_age_or_prek
borough                       ...                    
all boroughs             428  ...                 428
bronx                  70940  ...               70940
brooklyn               62231  ...               62231
connecticut              209  ...                 209
manhattan              68169  ...               68169
nassau county           4123  ...                4123
new jersey              1581  ...                1581
queens                 38056  ...               38056
rockland county          874  ...                 874
staten island          13906  ...               13906
westchester             7091  ...                7091

[11 rows x 20 columns]>

## 7. Follow up questions

### a). Did we have the right data?

The data provided helped address all our objectives

### b). Do we need other data to answer our question?

No

### c). Did we have the right question?

Yes